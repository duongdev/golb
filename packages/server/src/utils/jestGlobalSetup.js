import { MongoMemoryServer } from 'mongodb-memory-server'
import Debug from 'debug'
import * as dotenv from 'dotenv'

const NODE_ENV = process.env.NODE_ENV || 'development'

const fileList = [
  `.env.${NODE_ENV}.local`,
  `.env.${NODE_ENV}`,
  '.env.local',
  '.env',
]

// This util downloads the binary of MongoMemoryServer before the tests runs...
module.exports = async () => {
  const debug = Debug('globalSetup')

  fileList.forEach((file) => {
    dotenv.config({ path: file })
  })

  const mongoServer = new MongoMemoryServer()
  await mongoServer.getConnectionString()

  await mongoServer.stop()
  debug('Initialized MongoDB Memory Server successfully.')
}
