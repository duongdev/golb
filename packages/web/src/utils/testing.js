// Mocks useRouter
const useRouter = jest.spyOn(require('next/router'), 'useRouter')

/**
 * mockNextUseRouter
 * Mocks the useRouter React hook from Next.js on a test-case by test-case basis
 */

export function mockNextUseRouter(
  { route, pathname, query, asPath },
  { once = false } = {},
) {
  const router = {
    route,
    pathname,
    query,
    asPath,
    replace: jest.fn(),
  }
  useRouter[once ? 'mockImplementationOnce' : 'mockImplementation'](
    () => router,
  )
  return router
}
