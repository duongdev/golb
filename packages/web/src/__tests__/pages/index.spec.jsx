import React from 'react'
import { render, screen } from '@testing-library/react'
import { mockNextUseRouter } from 'utils/testing'
import App from '../../pages/index'

describe('pages/index', () => {
  const data = {
    docs: [],
    totalDocs: 0,
    limit: 10,
    totalPages: 1,
    page: 1,
    pagingCounter: 1,
    hasPrevPage: false,
    hasNextPage: false,
    prevPage: null,
    nextPage: null,
  }

  beforeEach(() => {
    mockNextUseRouter({
      route: '/',
      pathname: '/',
      query: '',
      asPath: `/`,
    })
  })
  afterAll(() => {
    jest.clearAllMocks()
    jest.restoreAllMocks()
  })
  it('renders correct message when the post list is empty', () => {
    render(<App data={data} />)
    expect(screen.getByTestId('empty-message').innerHTML).toMatchInlineSnapshot(
      `"There is no posts created yet."`,
    )
  })

  it('renders correct message when the search result is empty', () => {
    render(
      <App
        data={{
          ...data,
          searchText: 'something does not exist',
        }}
      />,
    )
    expect(screen.getByTestId('empty-message').innerHTML).toMatchInlineSnapshot(
      `"Couldn't find any posts relevant to <strong>something does not exist</strong>."`,
    )
  })

  it(`doesn't render the Pagination if there's only 1 page`, () => {
    render(<App data={data} />)
    expect(screen.queryByTestId('pagination')).not.toBeInTheDocument()
  })

  it(`renders the Pagination if there're 2 pages`, () => {
    render(<App data={{ ...data, totalPages: 2 }} />)
    expect(screen.queryByTestId('pagination')).toBeInTheDocument()
    expect(screen.queryByTestId('pagination').innerHTML).toMatchSnapshot()
  })

  it(`renders the list of posts correctly`, () => {
    render(
      <App
        data={{
          ...data,
          totalPages: 1,
          docs: [
            {
              id: 1,
              title: 'The first post',
              plainText: 'The first post content',
            },
            {
              id: 2,
              title: 'The second post',
              plainText: 'The second post content',
            },
          ],
        }}
      />,
    )
    expect(screen.queryAllByTestId('post-grid-item').length).toEqual(2)
  })
})
