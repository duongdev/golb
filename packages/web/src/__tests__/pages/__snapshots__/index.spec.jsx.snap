// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/index renders the Pagination if there're 2 pages 1`] = `"<ul class=\\"MuiPagination-ul\\"><li><a class=\\"MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-textPrimary Mui-disabled Mui-disabled\\" tabindex=\\"-1\\" aria-disabled=\\"true\\" aria-label=\\"Go to previous page\\" href=\\"/?page=0\\"><svg class=\\"MuiSvgIcon-root MuiPaginationItem-icon\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z\\"></path></svg></a></li><li><a class=\\"MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-textPrimary Mui-selected\\" tabindex=\\"0\\" aria-disabled=\\"false\\" aria-current=\\"true\\" aria-label=\\"page 1\\" href=\\"/?page=1\\">1<span class=\\"MuiTouchRipple-root\\"></span></a></li><li><a class=\\"MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-textPrimary\\" tabindex=\\"0\\" aria-disabled=\\"false\\" aria-label=\\"Go to page 2\\" href=\\"/?page=2\\">2<span class=\\"MuiTouchRipple-root\\"></span></a></li><li><a class=\\"MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-textPrimary\\" tabindex=\\"0\\" aria-disabled=\\"false\\" aria-label=\\"Go to next page\\" href=\\"/?page=2\\"><svg class=\\"MuiSvgIcon-root MuiPaginationItem-icon\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z\\"></path></svg><span class=\\"MuiTouchRipple-root\\"></span></a></li></ul>"`;
